[DPE-5094] tls implementation

MiaAltieri opened the pull request:
## Issue
We have no guarantees that TLS works on mongos router

## Solution
Add integration tests to guarantee that TLS works on mongos router + necessary changes to charm

MiaAltieri committed:
WIP TLS tests

MiaAltieri committed:
all tls tests work expect mismatch CA

MiaAltieri committed:
all tls tests are working

MiaAltieri committed:
fmt + lint

MiaAltieri committed:
use wait for blocked in tests

Gu1nness commented on .gitignore:
```diff
@@ -1,6 +1,7 @@
 venv/
 build/
 *.charm
+*.crt
```
This is bad workflow to ignore the files created in the test.

It would be way better to put them in a [TemporaryDirectory](https://docs.python.org/3/library/tempfile.html#tempfile.TemporaryDirectory) that will be fully when going out of the context manager.
Or event directly a `TemporaryFile`.
I can help with that if you want :)

MiaAltieri replied:
that would be a huge help, yes please 🙇‍♀️

Gu1nness replied:
Ok, do you want me to commit on your branch or should we talk about it during our 1-1?

MiaAltieri replied:
please commit directly! :)

Gu1nness replied:
Ok :)
As I thought, with pytest it's even easier since it provides a fixture just for that.

Gu1nness replied:
Done

Gu1nness commented on src/charm.py:
```diff
@@ -133,6 +134,18 @@ def _on_update_status(self, _):
             )
             return
 
+        if self.cluster.get_tls_statuses():
+            self.status.set_and_share_status(self.cluster.get_tls_statuses())
```
Use ternary operator here:
```suggestion
        if (tls_statuses := self.cluster.get_tls_statuses()):
            self.status.set_and_share_status(tls_statuses)
```

Gu1nness commented on tests/integration/tls/__init__.py:
```diff
@@ -0,0 +1,2 @@
+# Copyright 2023 Canonical Ltd.
```
```suggestion
# Copyright 2024 Canonical Ltd.
```

MiaAltieri committed:
Neha's suggestions from code review

Co-authored-by: Neha Oudin <17551419+Gu1nness@users.noreply.github.com>

Gu1nness committed:
chore: Use tmp path fixture

In order to avoid files left in the working directory, use the tmp_path
fixture from pytest to have temporary directories for this run.

Mehdi-Bendriss reviewed (changes requested):
Can you confirm that `_get_mongos_config_for_user` is still needed? If yes - we need to adjust it for TLS, if not - let's drop it

Mehdi-Bendriss commented on src/charm.py:
```diff
@@ -424,6 +490,14 @@ def unit_host(self, unit: Unit) -> str:
         unit_id = unit.name.split("/")[1]
         return f"{self.app.name}-{unit_id}.{self.app.name}-endpoints"
 
+    def get_config_server_name(self) -> Optional[str]:
```
is this really needed? we could just call `cluster.get_config_server_name`

MiaAltieri replied:
Yes - TLS lib makes call to `self.charm.get_config_server_name()` this is meant to be abstracted for:
- mongos vm 
- mongos k8s
- shards integrated to config server

MiaAltieri committed:
remove unused function

MiaAltieri committed:
Merge branch '6/edge' into DPE-5094-tls-implementation
